File: src/ObjectManager/Object.Tes/FilePacks/Records/345-FACT.Faction.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class FACTRecord : Record
    {
        // TESX
        public class RNAMGroup
        {
            public override string ToString() => $"{RNAM.Value}:{MNAM.Value}";
            public IN32Field RNAM; // rank
            public STRVField MNAM; // male
            public STRVField FNAM; // female
            public STRVField INAM; // insignia
        }

        // TES3
        public struct FADTField
        {
            public FADTField(UnityBinaryReader r, int dataSize)
            {
                r.SkipBytes(dataSize);
            }
        }

        // TES4
        public struct XNAMField
        {
            public override string ToString() => $"{FormId}";
            public int FormId;
            public int Mod;
            public int Combat;

            public XNAMField(UnityBinaryReader r, int dataSize, GameFormatId format)
            {
                FormId = r.ReadLEInt32();
                Mod = r.ReadLEInt32();
                Combat = format > GameFormatId.TES4 ? r.ReadLEInt32() : 0; // 0 - Neutral, 1 - Enemy, 2 - Ally, 3 - Friend
            }
        }

        public override string ToString() => $"FACT: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public STRVField FNAM; // Faction name
        public List<RNAMGroup> RNAMs = new List<RNAMGroup>(); // Rank Name
        public FADTField FADT; // Faction data
        public List<STRVField> ANAMs = new List<STRVField>(); // Faction name
        public List<INTVField> INTVs = new List<INTVField>(); // Faction reaction
        // TES4
        public XNAMField XNAM; // Interfaction Relations
        public INTVField DATA; // Flags (byte, uint32)
        public UI32Field CNAM;

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            if (format == GameFormatId.TES3)
                switch (type)
                {
                    case "NAME": EDID = r.ReadSTRV(dataSize); return true;
                    case "FNAM": FNAM = r.ReadSTRV(dataSize); return true;
                    case "RNAM": RNAMs.Add(new RNAMGroup { MNAM = r.ReadSTRV(dataSize) }); return true;
                    case "FADT": FADT = new FADTField(r, dataSize); return true;
                    case "ANAM": ANAMs.Add(r.ReadSTRV(dataSize)); return true;
                    case "INTV": INTVs.Add(r.ReadINTV(dataSize)); return true;
                    default: return false;
                }
            switch (type)
            {
                case "EDID": EDID = r.ReadSTRV(dataSize); return true;
                case "FULL": FNAM = r.ReadSTRV(dataSize); return true;
                case "XNAM": XNAM = new XNAMField(r, dataSize, format); return true;
                case "DATA": DATA = r.ReadINTV(dataSize); return true;
                case "CNAM": CNAM = r.ReadT<UI32Field>(dataSize); return true;
                case "RNAM": RNAMs.Add(new RNAMGroup { RNAM = r.ReadT<IN32Field>(dataSize) }); return true;
                case "MNAM": RNAMs.Last().MNAM = r.ReadSTRV(dataSize); return true;
                case "FNAM": RNAMs.Last().FNAM = r.ReadSTRV(dataSize); return true;
                case "INAM": RNAMs.Last().INAM = r.ReadSTRV(dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-FACT.Faction.swift
//
//  FACTRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class FACTRecord: Record {
    // TESX
    public class RNAMGroup: CustomStringConvertible {
        public var description: String { return "FACT: \(RNAM!):\(MNAM ?? "")" }
        public var RNAM: IN32Field! // rank
        public var MNAM: STRVField! // male
        public var FNAM: STRVField! // female
        public var INAM: STRVField! // insignia
        
        init(MNAM: STRVField) { self.MNAM = MNAM }
        init(RNAM: IN32Field) { self.RNAM = RNAM }
    }

    // TES3
    public struct FADTField {
        init(_ r: BinaryReader, _ dataSize: Int) {
            r.skipBytes(dataSize)
        }
    }

    // TES4
    public typealias XNAMField = (
        formId: Int32,
        mod: Int32,
        // TES5
        combat: Int32 // 0 - Neutral, 1 - Enemy, 2 - Ally, 3 - Friend
    )

    public override var description: String { return "FACT: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var FNAM: STRVField! // Faction name
    public var RNAMs = [RNAMGroup]() // Rank Name
    public var FADT: FADTField! // Faction data
    public var ANAMs = [STRVField]() // Faction name
    public var INTVs = [INTVField]() // Faction reaction
    // TES4
    public var XNAM: XNAMField! // Interfaction Relations
    public var DATA: INTVField! // Flags (byte, uint32)
    public var CNAM: UI32Field!

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        if format == .TES3 {
            switch type {
            case "NAME": EDID = r.readSTRV(dataSize)
            case "FNAM": FNAM = r.readSTRV(dataSize)
            case "RNAM": RNAMs.append(RNAMGroup(MNAM: r.readSTRV(dataSize)))
            case "FADT": FADT = FADTField(r, dataSize)
            case "ANAM": ANAMs.append(r.readSTRV(dataSize))
            case "INTV": INTVs.append(r.readINTV(dataSize))
            default: return false
            }
            return true
        }
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "FULL": FNAM = r.readSTRV(dataSize)
        case "XNAM": XNAM = r.readT(dataSize)
        case "DATA": DATA = r.readINTV(dataSize)
        case "CNAM": CNAM = r.readT(dataSize)
        case "RNAM": RNAMs.append(RNAMGroup(RNAM: r.readT(dataSize)))
        case "MNAM": RNAMs.last!.MNAM = r.readSTRV(dataSize)
        case "FNAM": RNAMs.last!.FNAM = r.readSTRV(dataSize)
        case "INAM": RNAMs.last!.INAM = r.readSTRV(dataSize)
        default: return false
        }
        return true
    }
}
